### PR TITLE
WIP: Doc Transition

### PR DESCRIPTION
**Pull request by MikeInnes**
This moves our 1503 docstrings from helpdb.jl into the new doc system. Docstrings are now stored in `base/docs/helpdb.jl` (soon to be scattered around base) and we have a hackish script which dumps those docstrings into the RST manual (at least until we move to markdown).

Caveats:
- Undocumented globals: `["ans", "CPU_CORES", "JULIA_HOME", "STDOUT", "STDERR", "STDIN"]` – need to implement docs for variable bindings to fix this.
- helpdb didn't support multimethods well, so some functions and docstrings have duplicated entries.

On the whole though we should only have lost on the order of tens of docstrings, as opposed to thousands.

### CODE REVIEW

**Comment by MichaelHatherly** [comment]
> unless you really hate blue

Could we just store them as plain text rather than ``MD`` objects?

**Comment by ViralBShah** [comment]
@MichaelHatherly Are you coming to JuliaCon? Would be great to sort this all out in person in case you are planning to be at MIT.

**Comment by MichaelHatherly** [comment]
@ViralBShah unfortunately I live many, many timezones away :( I'd have loved to be there though. I should be around for most of tomorrow so will be able to follow along.

**Comment by one-more-minute** [comment]
> Could we just store them as plain text rather than MD objects?

Technically yes, although there's a risk of setting a precedent of plain text docs in Base being acceptable.  This way it's obvious that the RST docs are a work in progress, which might be a good thing.

**Comment by MichaelHatherly** [comment]
> risk of setting the precedent of plain text docs in Base being acceptable

Hopefully not if potential additions are reviewed, but I'm not particularly bothered about blue text for a little while &ndash; might make things move a bit faster :)

**Comment by hayd** [comment]
I had a branch which did the rst to md conversion (or pretty close) to something like https://gist.github.com/hayd/9cbd69e5e85693d1f6eb

I think it's this: https://github.com/hayd/julia/commit/246314b2e5518db6c1960c7ba1b693649b9c51d3

Uses pycall/pypandoc to convert the rst docstrings to markdown, they need a little manual tweaking but pypandoc does a pretty good job.

File: doc/genstdlib.jl [hunk 1]
```diff
@@ -0,0 +1,80 @@
+using .Markdown
```
Comment (hayd) [inline doc/genstdlib.jl]: This file is missing a header, as well as using 2 space indentation (rather than 4).

File: doc/stdlib/arrays.rst [hunk 2]
```diff
@@ -80,117 +51,134 @@ largest range along each dimension.
...
 
-.. function:: eye(m, n)
+   Constructs an identity matrix of the same dimensions and type as
 
-   m-by-n identity matrix
 
 .. function:: eye(A)
 
-   Constructs an identity matrix of the same dimensions and type as ``A``.
+   Constructs an identity matrix of the same dimensions and type as
```
Comment (hayd) [inline doc/stdlib/arrays.rst]: This appears to be missing the remainder of the line (after a \`\`... not sure if what's the pattern, but there's quite a few lines below which demonstrate similar - the diff should definitely be combed through).

Also this (`function eye(A)`) now appears three times as the same signature :s

File: doc/stdlib/arrays.rst [hunk 3]
```diff
@@ -80,117 +51,134 @@ largest range along each dimension.
...
 
 .. function:: linspace(start, stop, n=100)
 
    Construct a range of ``n`` linearly spaced elements from ``start`` to ``stop``.
 
+
 .. function:: logspace(start, stop, n=50)
 
-   Construct a vector of ``n`` logarithmically spaced numbers from ``10^start`` to ``10^stop``.
+   Construct a vector of ``n`` logarithmically spaced numbers from
```
Comment (hayd) [inline doc/stdlib/arrays.rst]: As another example, this is missing the second line here?

File: doc/stdlib/arrays.rst [hunk 4]
```diff
@@ -199,490 +187,486 @@ All mathematical operations and functions are supported for arrays
 
 .. function:: broadcast(f, As...)
 
-   Broadcasts the arrays ``As`` to a common size by expanding singleton dimensions, and returns an array of the results ``f(as...)`` for each position.
+   Broadcasts the arrays ``As`` to a common size by expanding singleton dimensions, and returns an array of the results
```
Comment (hayd) [inline doc/stdlib/arrays.rst]: same here.

File: doc/genstdlib.jl (hunk 1 above)
Comment (tkelman) [inline doc/genstdlib.jl]: ref https://github.com/one-more-minute/Julia-LT/issues/167

**Comment by one-more-minute** [comment]
https://gist.github.com/one-more-minute/c4504665282016ec820b

Lines that are missing have the regex `-   [^\w ]`

**Comment by one-more-minute** [comment]
Ok, so the diff in the manual is still a little messy but I think it's mostly due to method definitions clobbering each other. e.g. in the new system we have
```julia
doc"""
    length(A) -> Integer

Returns the number of elements in A
"""
length

doc"""
    length(s)

The number of characters in string "s".
"""
length
```
The second clobbers the first so you get two (or more) copies of the second in the manual.

This isn't as bad as it sounds, because when the docstrings are attached to the right methods in Base they'll combine together and we'll get the whole thing back.

**Comment by StefanKarpinski** [comment]
Superseded by https://github.com/JuliaLang/julia/pull/11906.

**Comment by Ismael-VC** [comment]
> Undocumented globals: ["ans", "CPU_CORES", "JULIA_HOME", "STDOUT", "STDERR", "STDIN"] – need to implement docs for variable bindings to fix this.

One of the few things I liked about .NET when I was forced to use Visual Studio (and that now I miss in XCode) was the ability to document global variables:

```csharp
/// <summary>
/// Always use this
/// </summary>
List<string> my_list = List<string>();
```

Could we do something like this work?

```julia
"Returns the number of available CPU cores."
global CPU_CORES

function init_sysinfo()
    # set CPU core count
    global const CPU_CORES =
        haskey(ENV,"JULIA_CPU_CORES") ? parse(Int,ENV["JULIA_CPU_CORES"]) :
                                        Int(ccall(:jl_cpu_cores, Int32, ()))
    global const SC_CLK_TCK = ccall(:jl_SC_CLK_TCK, Clong, ())
    global const cpu_name = ccall(:jl_get_cpu_name, ByteString, ())
end
```

**Comment by one-more-minute** [comment]
Yeah it's totally doable, just needs implementing.

**Pull request closed without merging**